src/number_theory/diophantine_approximation.lean

```diff
@@ -265,3 +284,317 @@ begin
+        ← add_lt_add_iff_right (v * (2 * v - 1) : ℝ), add_comm (1 : ℝ)] at h,
+    have := (mul_lt_mul_right $ hv₀').mpr ((sub_lt_sub_iff_left (u : ℝ)).mpr $
+               (mul_lt_mul_right hv₀).mpr $ sub_right_lt_of_lt_add $ lt_floor_add_one ξ),
+    rw [sub_mul ξ, one_mul, ← sub_add, add_mul] at this,
+    exact_mod_cast this.trans h, },
+  have huv_cop : is_coprime (u - ⌊ξ⌋ * v) v,
+  { rwa [sub_eq_add_neg, ← neg_mul, is_coprime.add_mul_right_left_iff], },
+  refine ⟨lt_of_le_of_ne' hu₀ (λ hf, _), lt_of_le_of_ne hu₁ (λ hf, _)⟩;
+  { rw hf at huv_cop,
+    simp only [is_coprime_zero_left, is_coprime_self, is_unit_iff] at huv_cop,
+    cases huv_cop; linarith only [hv, huv_cop], },
+end
+
+/-- The key step: the relevant inequality persists in the inductive step. -/
+lemma aux₃ : |1 / fract ξ - v / (u - ⌊ξ⌋ * v)| < 1 / ((u - ⌊ξ⌋ * v) * (2 * (u - ⌊ξ⌋ * v) - 1)) :=
+begin
+  obtain ⟨hu₀, huv⟩ := aux₂ hv h,
+  have hξ₀ := aux₁ hv h,
+  set u' := u - ⌊ξ⌋ * v with hu',
+  have hu'ℝ : (u' : ℝ) = u - ⌊ξ⌋ * v := by exact_mod_cast hu',
+  rw ← hu'ℝ,
+  replace hu'ℝ := (eq_sub_iff_add_eq.mp hu'ℝ).symm,
+  obtain ⟨Hu, Hu'⟩ := aux₀ hu₀,
+  obtain ⟨Hv, Hv'⟩ := aux₀ (zero_lt_two.trans_le hv),
+  have H₁ := div_pos (div_pos Hv Hu) hξ₀,
+  replace h := h.2.2,
+  have h' : |fract ξ - u' / v| < 1 / (v * (2 * v - 1)),
+  { rwa [hu'ℝ, add_div, mul_div_cancel _ Hv.ne', ← sub_sub, sub_right_comm] at h, },
+  have H : (2 * u' - 1 : ℝ) ≤ (2 * v - 1) * fract ξ,
+  { replace h := (abs_lt.mp h).1,
+    have : (2 * (v : ℝ) - 1) * (-(1 / (v * (2 * v - 1))) + u' / v) = 2 * u' - (1 + u') / v,
+    { field_simp [Hv.ne', Hv'.ne'], ring, },
+    rw [hu'ℝ, add_div, mul_div_cancel _ Hv.ne', ← sub_sub, sub_right_comm, self_sub_floor,
+        lt_sub_iff_add_lt, ← mul_lt_mul_left Hv', this] at h,
+    refine has_le.le.trans _ h.le,
+    rw [sub_le_sub_iff_left, div_le_one Hv, add_comm],
+    exact_mod_cast huv, },
+  have help₁ : ∀ {a b c : ℝ}, a ≠ 0 → b ≠ 0 → c ≠ 0 →
+                 |1 / a - b / c| = |(a - c / b) * (b / c / a)|,
+  { intros, rw abs_sub_comm, congr' 1, field_simp, ring },
+  have help₂ : ∀ {a b c d : ℝ}, a ≠ 0 → b ≠ 0 → c ≠ 0 → d ≠ 0 →
+                 1 / (b * c) * (b / d / a) = 1 / (d * c * a),
+  { intros, field_simp, ring },
+  calc
+    |1 / fract ξ - v / u'|
+        = |(fract ξ - u' / v) * (v / u' / fract ξ)| : help₁ hξ₀.ne' Hv.ne' Hu.ne'
+    ... = |fract ξ - u' / v| * (v / u' / fract ξ) :   by rw [abs_mul, abs_of_pos H₁, abs_sub_comm]
+    ... < 1 / (v * (2 * v - 1)) * (v / u' / fract ξ) : (mul_lt_mul_right H₁).mpr h'
+    ... = 1 / (u' * (2 * v - 1) * fract ξ) :          help₂ hξ₀.ne' Hv.ne' Hv'.ne' Hu.ne'
+    ... ≤ 1 / (u' * (2 * u' - 1)) : by rwa [one_div_le_one_div (mul_pos (mul_pos Hu Hv') hξ₀) $
+                                              mul_pos Hu Hu', mul_assoc, mul_le_mul_left Hu],
+end
+
+/-- The conditions `ass ξ u v` persist in the inductive step. -/
+lemma invariant : ass (1 / fract ξ) v (u - ⌊ξ⌋ * v) :=
```

Comment:
This is a pretty uninformative name. Can you improve it or mark this lemma as private (you'll have to get rid of the section end)?

Reply:
I'm marking it as private; I don't think it is useful outside the proof of the main result.